### PR TITLE
BUG-FIX - Check count and bytes threshold on every insert into to buffer

### PR DESCRIPTION
**Pull request by sfc-gh-japatel**
- Refactored buffer threshold logic inside another final class
- Fixed bug where thresholds were not checked on every records insertion
  to buffer.

### CODE REVIEW

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java [hunk 1]
```diff
@@ -317,39 +343,68 @@ private SinkRecord handleNativeRecord(SinkRecord record, boolean isKey) {
...
+   * <p>Threshold is config parameter: {@link
+   * com.snowflake.kafka.connector.SnowflakeSinkConnectorConfig#BUFFER_FLUSH_TIME_SEC}
+   *
+   * <p>Previous flush time here means last time we called insertRows API with rows present in
+   */
+  protected void insertBufferedRowsIfFlushTimeThresholdReached() {
+    if (this.streamingBufferThreshold.isFlushTimeBased(this.previousFlushTimeStampMs)) {
+      LOGGER.debug(
+          "Time based flush for channel:{}, CurrentTimeMs:{}, previousFlushTimeMs:{},"
+              + " bufferThresholdMs:{}",
```
Comment (sfc-gh-japatel) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:362]: Fill fix this

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java [hunk 2]
```diff
@@ -186,22 +190,14 @@ public void insert(Collection<SinkRecord> records) {
...
-          >= (getFlushTime() * 1000)) {
-        LOGGER.info("Time based flush for channel:{}", partitionChannel.getChannelName());
-        LOGGER.info(
-            "Current:{}, previousFlushTime:{}, threshold:{}",
-            System.currentTimeMillis(),
-            partitionChannel.getPreviousFlushTimeStampMs(),
-            (getFlushTime() * 1000));
-        partitionChannel.insertBufferedRows();
-      }
+      partitionChannel.insertBufferedRowsIfFlushTimeThresholdReached();
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java:200]: I believe we have a TODO somewhere to make this check happens in a separate thread?

File: src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java [hunk 3]
```diff
@@ -317,39 +343,68 @@ private SinkRecord handleNativeRecord(SinkRecord record, boolean isKey) {
...
+          this.streamingBufferThreshold.getFlushTimeThresholdSeconds());
+      StreamingBuffer copiedStreamingBuffer;
+      bufferLock.lock();
+      try {
+        copiedStreamingBuffer = this.streamingBuffer;
+        this.streamingBuffer = new StreamingBuffer();
+      } finally {
+        bufferLock.unlock();
+      }
+      insertBufferedRows(copiedStreamingBuffer);
```
Comment (sfc-gh-tzhang) [inline src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java:375]: Not sure if you need to check if `copiedStreamingBuffer` is null (at least you did it in line 254

**Review by sfc-gh-tzhang — APPROVED** [review]
LGTM!

**Pull request merged**